### PR TITLE
Reward Program TLS v2 communication

### PR DESCRIPTION
**Pull request by fboucquez**
Related to https://github.com/nemgrouplimited/symbol-node-rewards/pull/234

Generating agent certificates according to the new TLS communication. AgentCertificateService is based on the TLS PR's script.
https://github.com/nemgrouplimited/symbol-node-rewards/blob/f983484fcb900b2ce9c0da0fc1c857429e32947f/cmds/controller-certs/generate.sh

The agent-ca.csr.pem can be regenerated when using the original private key. This means that if the node admin loses the target folder, the file can be regenerated from a backed-up private key (similar to the node certificates) without requiring re-enrolment. 

agent.properties file has been updated, 

Updated healthCheck adding /medatata endoint check

This PR also includes the `enrol` to `enroll` rename.

### CODE REVIEW

**Comment by segfaultxavi** [comment]
This is in line with the US English we're trying to use everywhere 👍

File: README.md [hunk 1]
```diff
@@ -268,7 +268,7 @@ General users should install this tool like any other node module.
 * [`symbol-bootstrap config`](docs/config.md) - Command used to set up the configuration files and the nemesis block for the current network
 * [`symbol-bootstrap decrypt`](docs/decrypt.md) - It decrypts a yml file using the provided password. The source file can be a custom preset file, a preset.yml file or an addresses.yml.
 * [`symbol-bootstrap encrypt`](docs/encrypt.md) - It encrypts a yml file using the provided password. The source files would be a custom preset file, a preset.yml file or an addresses.yml.
-* [`symbol-bootstrap enrolRewardProgram`](docs/enrolRewardProgram.md) - It enrols the nodes in the rewards program by announcing the enrol transaction to the registration address.  You can also use this command to update the program registration when you change the node public key or server host.
+* [`symbol-bootstrap enrollRewardProgram`](docs/enrollRewardProgram.md) - It enrols the nodes in the rewards program by announcing the enroll transaction to the registration address.  You can also use this command to update the program registration when you change the node public key or server host.
```
Comment (yilmazbahadir) [inline README.md:271]: We should remove `when you change the node public key` part here. It doesn't take place in the enrollment message anymore and also it's not used in the Controller. Should we mention the `agent-ca-csr` here?

File: docs/enrollRewardProgram.md [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
 Currently, the only program that can be enrolled post-launch is 'SuperNode'.
 
-* [`symbol-bootstrap enrolRewardProgram`](#symbol-bootstrap-enrolrewardprogram)
+* [`symbol-bootstrap enrollRewardProgram`](#symbol-bootstrap-enrollrewardprogram)
 
-## `symbol-bootstrap enrolRewardProgram`
+## `symbol-bootstrap enrollRewardProgram`
 
-It enrols the nodes in the rewards program by announcing the enrol transaction to the registration address.  You can also use this command to update the program registration when you change the node public key or server host.
+It enrols the nodes in the rewards program by announcing the enroll transaction to the registration address.  You can also use this command to update the program registration when you change the node public key or server host.
```
Comment (yilmazbahadir) [inline docs/enrollRewardProgram.md:12]: remove `when you change the node public key` ?

File: src/service/RewardProgramService.ts [hunk 3]
```diff
@@ -119,7 +119,7 @@ export class RewardProgramService implements TransactionFactory {
             return transactions;
         }
         const agentUrl = nodePreset.agentUrl || `https://${nodePreset.host}:7880`;
```
Comment (yilmazbahadir) [inline src/service/RewardProgramService.ts:121]: The default port will be changed to `7881` instead (with this PR => https://github.com/nemgrouplimited/symbol-node-rewards/pull/234)

File: src/service/AgentCertificateService.ts [hunk 4]
```diff
@@ -53,20 +95,74 @@ export class AgentCertificateService {
...
+chmod 700 new_certs
+touch index.txt
+
 # Creating a self signed certificate for the agent. This will be created by bootstrap when setting up a supernode
-openssl genrsa -out agent-key.pem 4096
-openssl req -new -config agent.cnf -key agent-key.pem -out agent-csr.pem
-openssl x509 -req -extfile agent.cnf -days 999 -in agent-csr.pem -out agent-crt.pem -signkey agent-key.pem
+
+cat agent-ca.der | openssl pkey -inform DER -outform PEM -out agent-ca.key.pem
+openssl pkey -inform pem -in agent-ca.key.pem -text -noout
```
Comment (yilmazbahadir) [inline src/service/AgentCertificateService.ts:158]: just to make sure, `agent-ca.der` is generated using `ed25519` algorithm, right?

File: test/service/RewardProgramService.test.ts [hunk 5]
```diff
@@ -66,7 +66,7 @@ describe('RewardProgramService', () => {
         const transaction = transactions[0] as TransferTransaction;
         expect(transaction.type).eq(TransactionType.TRANSFER);
         expect(transaction.message.payload).eq(
-            'enrol C9767496987222790518114049299DD52114BF2A8F7E5F4B70BB2B6365FAFD34 https://fboucquez-agent-symbollocal.ngrok.io',
+            'enroll C9767496987222790518114049299DD52114BF2A8F7E5F4B70BB2B6365FAFD34 https://fboucquez-agent-symbollocal.ngrok.io',
```
Comment (yilmazbahadir) [inline test/service/RewardProgramService.test.ts:69]: Message should be in this format => `enroll <agent_url> <base64_agent_ca_csr_file>`

base64_agent_ca_csr_file => when decoded the text should include `-----BEGIN CERTIFICATE REQUEST---xxx -----END CERTIFICATE REQUEST-----`

File: src/service/RewardProgramService.ts [hunk 6]
```diff
@@ -119,7 +119,7 @@ export class RewardProgramService implements TransactionFactory {
...
-        const plainMessage = `enrol ${agentPublicKey} ${agentUrl}`;
+        const plainMessage = `enroll ${agentPublicKey} ${agentUrl}`;
```
Comment (yilmazbahadir) [inline src/service/RewardProgramService.ts:122]: Message should be in this format => `enroll <agent_url> <base64_agent_ca_csr_file>`

base64_agent_ca_csr_file => when decoded the text should include
` -----BEGIN CERTIFICATE REQUEST---xxx -----END CERTIFICATE REQUEST-----`

File: src/service/AgentCertificateService.ts (hunk 4 above)
Comment (fboucquez) [reply src/service/AgentCertificateService.ts:158]: Yes, it also check the openssl private and public keys against the sdk generated ones

File: test/service/RewardProgramService.test.ts (hunk 5 above)
Comment (fboucquez) [reply test/service/RewardProgramService.test.ts]: Updated

File: docs/enrollRewardProgram.md [hunk 7]
```diff
@@ -1,19 +1,19 @@
-`symbol-bootstrap enrolRewardProgram`
-=====================================
+`symbol-bootstrap enrollRewardProgram`
+======================================
 
-It enrols the nodes in the rewards program by announcing the enrol transaction to the registration address.  You can also use this command to update the program registration when you change the node public key or server host.
+It enrols the nodes in the rewards program by announcing the enroll transaction to the registration address.  You can also use this command to update the program registration when you change the agent keys (changing the agent-ca-csr) or server host.
 
 Currently, the only program that can be enrolled post-launch is 'SuperNode'.
```
Comment (yilmazbahadir) [inline docs/enrollRewardProgram.md:6]: We should discuss how the 'Ecosystem' and 'EarlyAdoption' node owners will enroll with the new agentCaCsr? the only place is Nis1 for now, I guess.

File: config/agent/agent.properties.mustache [hunk 8]
```diff
@@ -1,8 +1,9 @@
-NODE_PRIVATE_KEY={{{nodePrivateKey}}}
-LOGGER_FILE=logs/agent.log
+CERTS_CONTROLLER_CA_CERT_FILE={{{rewardProgramCaFile}}}
+CONTROLLER_PUBLIC_KEY={{{rewardProgramControllerPublicKey}}}
```
[the remark this answers is not in the document]
Comment (fboucquez) [reply config/agent/agent.properties.mustache:2]: The address from the public key is used to send the enrol message

File: docs/enrollRewardProgram.md (hunk 7 above)
Comment (fboucquez) [reply docs/enrollRewardProgram.md:6]: One Idea is we populate the initial node entry in the collection with an empty certificate and we tell node admins to send the eroll message. Early and ecosystem can just update the node entry, not create

File: config/agent/agent.properties.mustache (hunk 8 above)
[the remark this answers is not in the document]
Comment (yilmazbahadir) [reply config/agent/agent.properties.mustache:2]: I can see `rewardProgramControllerPublicKey` is used for that purpose but couldn't find any reference to `CONTROLLER_PUBLIC_KEY` in the code(both bootstrap and the agent). Can you delete it from `agent.properties.mustache` if we don't use/need it?

[the remark this answers is not in the document]
Comment (fboucquez) [reply config/agent/agent.properties.mustache]: Thanks!
I've removed `CONTROLLER_PUBLIC_KEY` from `agent.properties.mustache` and I have moved from `rewardProgramControllerPublicKey` with `rewardProgramControllerAddress` so it easier.

[the remark this answers is not in the document]
Comment (yilmazbahadir) [reply config/agent/agent.properties.mustache]: Tbh, something like  `rewardProgramEnroll(ment)Address` makes more sense to me. Even in the controller we store it in `ENROLMENT_ADDRESS` (which has a spelling error which I'll correct').

**Review by yilmazbahadir — APPROVED** [review]

**Pull request merged**